File: Medium/74. Search a 2D Matrix/binary search 2d matrix.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    const rows = matrix.length;
    const cols = matrix[0].length;

    let l = 0;
    let r = rows * cols - 1;

    while (l <= r) {
        const mid = l + Math.floor((r - l) / 2);

        const row = Math.floor(mid / cols);
        const col = mid % cols;

        const val = matrix[row][col];

        if (val === target) {
            return true;
        } else if (val > target) {
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }

    return false;
}

File: Medium/74. Search a 2D Matrix/binary search 2d matrix.go
func searchMatrix(matrix [][]int, target int) bool {
	rows, cols := len(matrix), len(matrix[0])

	l, r := 0, rows*cols-1
	var m int

	for l <= r {
		m = l + (r-l)/2

		row, col := m/cols, m%cols

		val := matrix[row][col]

		if val == target {
			return true
		} else if val > target {
			r = m - 1
		} else {
			l = m + 1
		}
	}

	return false
}
